acpi: find all subtables usign foreach() helpers

finikorg opened the pull request:
At the moment only first `ACPI_DMAR_HARDWARE_UNIT` table is found, find all structures using foreach() helpers, verify with acpi info test for `up_squared_pro_7000`.

This can also fix issues with VTD (after later modifications) not found because IOAPIC device scope is located in not found `ACPI_DMAR_HARDWARE_UNIT` table.

Can be verified with `up_squared_pro_7000` and `intel_ehl_crb` board.

finikorg committed:
acpi: Add acpi_dmar_foreach helpers

Add function walking though all DMAR subtables, at the moment only
first subtable is taking into account, which causes bugs for some
boards.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: Refactor ACPI info sample

Take into account multiple DMAR HARDWARE_UNIT tables.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
boards: alder_lake: Enable VTD

Enable VTD for Alder Lake board.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -173,6 +173,11 @@ int acpi_dmar_entry_get(enum AcpiDmarType type, ACPI_SUBTABLE_HEADER **tables);
 int acpi_drhd_get(enum AcpiDmarScopeType scope, ACPI_DMAR_DEVICE_SCOPE *dev_scope,
 		  union acpi_dmar_id *dmar_id, int *num_inst, int max_inst);
 
+typedef void (*dmar_foreach_func_t)(uintptr_t subtable);
```
Why `uintptr_t` instead of e.g. `void *`? The previous `ACPI_DMAR_HEADER *` also seemed like a good approach, especially if `ACPI_DMAR_HEADER` is always the first element of each specialised DMAR struct - in that case you could use `CONTAINER_OF()` to get the actual type, without the need of any explicit or implicit type casts (which are always a bit risky in C since they will hide potential type conflicts).

finikorg committed:
drivers: intc_ioapic: Fix get ioapic_id

Information about IOAPIC can be located not in the first
DMAR Hardware Unit Definition subtable. Iterate them all.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -173,6 +173,11 @@ int acpi_dmar_entry_get(enum AcpiDmarType type, ACPI_SUBTABLE_HEADER **tables);
 int acpi_drhd_get(enum AcpiDmarScopeType scope, ACPI_DMAR_DEVICE_SCOPE *dev_scope,
 		  union acpi_dmar_id *dmar_id, int *num_inst, int max_inst);
 
+typedef void (*dmar_foreach_func_t)(void *subtable);
+
+void acpi_dmar_foreach_subtable(ACPI_TABLE_DMAR *dmar, dmar_foreach_func_t func);
+void acpi_dmar_foreach_devscope(ACPI_DMAR_HARDWARE_UNIT *hu, dmar_foreach_func_t func);
```
Does it really make sense to use the same callback type for both foreach functions? If you had a dedicated callback type, couldn't you then have the correct parameter type without the need for implicit casting from `void *`? I.e. one callback would have the input parameter type `ACPI_DMAR_DEVICE_SCOPE *` and the other would have `ACPI_DMAR_HEADER *`? Feel free to disagree - it's possible there's something I don't understand wrt to the ACPI typing here.

finikorg replied:
Using different callbacks, probably, makes code clearer, since I do not need to cast. I used the same as I though those structures have the same headers.

jhedberg commented on tests/arch/x86/info/src/acpi.c:
```diff
@@ -28,25 +24,33 @@ static const char *get_dmar_scope_type(int type)
 
-static void vtd_dev_scope_info(struct acpi_dmar_device_scope *dev_scope,
-			       union acpi_dmar_id *dmar_id, int num_inst)
+static void dmar_devsope_handler(uintptr_t arg)
```
Looks like you forgot to update the input parameter type here

jhedberg commented on tests/arch/x86/info/src/acpi.c:
```diff
@@ -60,25 +64,24 @@ static void vtd_drhd_info(struct acpi_dmar_hardware_unit *drhd)
+	acpi_dmar_foreach_devscope(drhd, dmar_devsope_handler);
+}
+
+static void dmar_subtable_handler(uintptr_t arg)
```
and here

jhedberg commented on drivers/interrupt_controller/intc_ioapic.c:
```diff
@@ -125,28 +125,64 @@ static const struct device *const vtd =
 
-	/* Assume only one PCH in system (say client platform). */
-	if (!acpi_drhd_get(ACPI_DMAR_SCOPE_TYPE_IOAPIC, NULL, &dmar_id, &inst_cnt, 1u)) {
+	acpi_dmar_foreach_devscope((void *)subtable, devscope_handler, arg);
```
Sorry to keep pressing you about the `void *` casting - I'm not against it when there's no simple way to avoid it, but if it can be avoided it would make it clearer what kind of types we're dealing with. In this case, I think the following would make the intention of the code more clear:
```c
ACPI_DMAR_HARDWARE_UNIT *hu;

hu = CONTAINER_OF(subtable, ACPI_DMAR_HARDWARE_UNIT, Header);

acpi_dmar_foreach_devscope(hu, devscop_handler, arg);
```

Maybe `hdr` (or similar) might be a more suitable variable name instead of `subtable` if you go with that approach.

tbursztyka commented on drivers/interrupt_controller/intc_ioapic.c:
```diff
@@ -125,28 +125,65 @@ static const struct device *const vtd =
+static bool get_ioapic(void)
+{
+	ACPI_TABLE_DMAR *dmar = acpi_table_get("DMAR", 0);
+
```
I don't like seeing all this ACPI related code into ioapic.

It should be minimal as possible. So move as much as possible into lib/acpi/

finikorg replied:
moved to acpi lib

tbursztyka commented on dts/x86/intel/alder_lake.dtsi:
```diff
@@ -258,6 +258,12 @@
 		compatible = "simple-bus";
 		ranges;
 
+		vtd: vtd@fed91000 {
```
Don't enable it yet. (move that commit away).
I understand you needed that to build-test vtd on that platform, but ADL does not expose any IOAPIC rerouting capability. (which either could mean we need to move forward a 100% PCIe based system or... or I don't see)
So vtd is not going to work yet on ADL because we still use too much legacy stuff atm.

tbursztyka commented on arch/x86/core/cache.c:
```diff
@@ -18,32 +18,40 @@
 #include <zephyr/cache.h>
 #include <stdbool.h>
 
+/* Not Write-through bit */
```
I'll push a white list on the relevant PR. remove that commit.

finikorg replied:
The commit would be rebased when you PR is merged. Without this I cannot compile and test the code.

tbursztyka reviewed (changes requested):
Verify vtd on EHL, not ADL.

finikorg committed:
lib: acpi: Implement acpi_dmar_ioapic_get()

Implement get IOAPIC id from DMAR table.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented on include/zephyr/acpi/acpi.h:
```diff
@@ -173,6 +173,22 @@ int acpi_dmar_entry_get(enum AcpiDmarType type, ACPI_SUBTABLE_HEADER **tables);
+ * @param ioapic_id IOAPIC id
+ * @return true if found false otherwise
+ */
+bool acpi_dmar_ioapic_get(uint16_t *ioapic_id);
```
I know this fits nicely into the existing `intc_ioapic.c` code, but should't we try to keep the API conventions consistent. You have other "get()" functions here (like acpi_drhd_get()) which uses the standard Zephyr int return style.

finikorg replied:
Fixed

finikorg commented:
Rebased to include cache fixes.

finikorg committed:
tests: acpi: Rename acpi to integration test

Rename test to integration and allow to include unit tests for ACPI.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
lib: acpi: Add ASSERT() for wrong length

Add ASSERT() for length.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: acpi: Add unit test for ACPI table parsing

Initial unit test for ACPI table parsing.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
tests: lib: acpi: Convert to FFF

Convert fake functions to FFF.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

jhedberg commented:
@finikorg it seems `tests/lib/acpi/boards` was left behind when you moved the sample to `tests/lib/acpi/integration`. Could either you or @najumon1980 submit a PR to fix that? I'm also surprised the sample still builds and runs with qemu_x86_64 even though the DTS overlay was left behind in that directory. Is it actually not needed?
